src/server/service/user.ts

```diff
@@ -30,7 +29,28 @@ export default class UserService implements IUserService {
     return hash;
   }
 
-  async create(dto: CreateUserDTO): Promise<User> {
+  async create(dto: CreateUserDTO, email: string): Promise<User> {
+    const existentUser = await this.findByEmail(email);
+    if (!dto.role) {
+      throw new Error('"role" is a required field!');
+    }
+
+    if (existentUser.role === 'admin') {
+      if (dto.role !== 'manager') {
+        throw new Error('Action not allowed admin role');
+      }
+    }
+
+    if (existentUser.role === 'manager') {
+      if (dto.role !== 'user') {
+        throw new Error('Action not allowed for manager role');
+      }
+    }
+
+    if (existentUser.role === 'user') {
+      throw new Error('Action not allowed for user role');
+    }
```

Comment:
I would also implemented an error for these cases, a single `ActionRequiresPromotion` error could help, this translates to an invalid action so we could also return a [422](https://httpstatuses.com/422) error.

Reply:
I would also change these 3 condition evaluations to use the `Role` enum.

```suggestion
    if (existentUser.role === Role.Admin) {
      if (dto.role !== 'manager') {
        throw new Error('Action not allowed admin role');
      }
    }

    if (existentUser.role === Role.Manager) {
      if (dto.role !== 'user') {
        throw new Error('Action not allowed for manager role');
      }
    }

    if (existentUser.role === Role.User) {
      throw new Error('Action not allowed for user role');
    }
```